### PR TITLE
Add a simple network as a supplementary source code for how to quantize document.

### PR DESCRIPTION
**Pull request by oatawa1**
## What this patch does to fix the issue.
Create a simple network as a supplementary for how to quantize document.

## Link to any relevant issues or pull requests.
#955

### CODE REVIEW

File: blueoil/networks/classification/quantize_example.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
+from functools import partial
+
+import tensorflow as tf
+
+from blueoil.blocks import darknet
+from blueoil.networks.classification.base import Base
+from blueoil.layers import conv2d
+
+
+class FooNetwork(Base):
```
Comment (iizukak) [inline blueoil/networks/classification/quantize_example.py:10]: IMO, `Foo` `Bar` is too casual to official sample code.
`SampleNetwork` is better.

File: blueoil/networks/classification/quantize_example.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    ):
+        super().__init__(
+            *args,
+            **kwargs,
+        )
+
+        self.activation = lambda x: tf.nn.leaky_relu(x, alpha=0.1, name="leaky_relu")
+        self.before_last_activation = self.activation
+
+    def base(self, images, is_training):
```
Comment (iizukak) [inline blueoil/networks/classification/quantize_example.py:26]: In example, we can assume just one data format.

File: blueoil/networks/classification/quantize_example.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+            channel_data_format = "channels_last"
+        else:
+            raise RuntimeError("data format {} should be in ['NCHW', 'NHWC]'.".format(self.data_format))
+
+        self.inputs = self.images = images
+
+        darknet_block = partial(darknet, is_training=is_training,
+                                activation=self.activation, data_format=self.data_format)
+
+        x = darknet_block("block_1", self.inputs, filters=32, kernel_size=1)
```
Comment (iizukak) [inline blueoil/networks/classification/quantize_example.py:39]: Is it possible not use `darknet_block` ?
Document reader dose not know about `darknet_block`.

Comment (oatawa1) [reply blueoil/networks/classification/quantize_example.py:39]: Now I figured out how to use the ops.

I think I will use blueoil layer(blueoil.layers.conv2d), thus the document reader will follow and use the blueoil layer. What do you think?

Comment (iizukak) [reply blueoil/networks/classification/quantize_example.py:39]: OK, I see.

File: blueoil/networks/classification/quantize_example.py (hunk 2 above)
Comment (oatawa1) [reply blueoil/networks/classification/quantize_example.py:25]: NHWC is choosed.

**Comment by oatawa1** [comment]
@iizukak Thanks for review and comments. I fixed them. Please review it again.

File: blueoil/networks/classification/quantize_example.py (hunk 2 above)
Comment (iizukak) [reply blueoil/networks/classification/quantize_example.py:25]: nice :)

File: blueoil/networks/classification/quantize_example.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+            *args,
+            **kwargs
+    ):
+        super().__init__(
+            *args,
+            **kwargs,
+        )
+
+        self.activation = lambda x: tf.nn.leaky_relu(x, alpha=0.1, name="leaky_relu")
+        self.before_last_activation = self.activation
```
Comment (iizukak) [inline blueoil/networks/classification/quantize_example.py:23]: What is the `before_last_activation` ?

File: blueoil/networks/classification/quantize_example.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+            **kwargs
+    ):
+        """
+        Args:
+            quantize_first_convolution(bool): use quantization in first conv.
+            quantize_last_convolution(bool): use quantization in last conv.
+            weight_quantizer (callable): weight quantizer.
+            weight_quantize_kwargs(dict): Initialize kwargs for weight quantizer.
+            activation_quantizer (callable): activation quantizer
+            activation_quantize_kwargs(dict): Initialize kwargs for activation quantizer.
```
Comment (iizukak) [inline blueoil/networks/classification/quantize_example.py:76]:
```suggestion
            activation_quantizer_kwargs(dict): Initialize kwargs for activation quantizer.
```

Comment (oatawa1) [reply blueoil/networks/classification/quantize_example.py:76]: I fixed not only this but also in others network as well (#967).

File: blueoil/networks/classification/quantize_example.py (hunk 4 above)
Comment (oatawa1) [reply blueoil/networks/classification/quantize_example.py:23]: Thank you for raise this up. 

I picked it from `darknet`. It used as a parameter to quantize activation function in the before last convolution block. 

I will fix and make use of it in this simple network.

Comment (oatawa1) [reply blueoil/networks/classification/quantize_example.py:23]: @iizukak Now, I keep it in the model to show how to quantize when `quantize_last_convolution=True`. I can remove it out though.

Do you think this is necessary?

**Comment by oatawa1** [comment]
@iizukak Could you review this PR again?

File: blueoil/networks/classification/quantize_example.py (hunk 4 above)
Comment (iizukak) [reply blueoil/networks/classification/quantize_example.py:23]: I think this is not needed for sample application.

Comment (oatawa1) [reply blueoil/networks/classification/quantize_example.py]: I removed this part out

**Comment by oatawa1** [comment]
> @oatawa1 Can you add unit test?

@iizukak I don't know how but I will try.

**Comment by iizukak** [comment]
@oatawa1 like
https://github.com/blue-oil/blueoil/blob/master/tests/unit/networks_tests/classification_test/test_lmnet_quantize.py

**Comment by oatawa1** [comment]
@iizukak Thanks, I added unit test.

**Comment by oatawa1** [comment]
@iizukak Thanks

File: blueoil/networks/classification/quantize_example.py [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+
+from blueoil.networks.classification.base import Base
+from blueoil.layers import batch_norm, conv2d
+
+
+class SampleNetwork(Base):
+    """Sample network with simple layer."""
+
+    def __init__(
+            self,
```
Comment (tsawada) [inline blueoil/networks/classification/quantize_example.py:13]: fits in a single line

File: blueoil/networks/classification/quantize_example.py [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+            *args,
+            **kwargs
+    ):
+        super().__init__(
```
Comment (tsawada) [inline blueoil/networks/classification/quantize_example.py:17]: fits in a single line

File: blueoil/networks/classification/quantize_example.py [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+        Args:
+            quantize_first_convolution(bool): use quantization in first conv.
+            quantize_last_convolution(bool): use quantization in last conv.
+            weight_quantizer (callable): weight quantizer.
+            weight_quantizer_kwargs(dict): Initialize kwargs for weight quantizer.
+            activation_quantizer (callable): activation quantizer
+            activation_quantizer_kwargs(dict): Initialize kwargs for activation quantizer.
+        """
+
+        super().__init__(
```
Comment (tsawada) [inline blueoil/networks/classification/quantize_example.py:77]: fits in a single line

File: blueoil/networks/classification/quantize_example.py [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+            *args,
+            **kwargs,
+        )
+
+        self.quantize_first_convolution = quantize_first_convolution
+        self.quantize_last_convolution = quantize_last_convolution
+
+        activation_quantizer_kwargs = activation_quantizer_kwargs if not None else {}
```
Comment (tsawada) [inline blueoil/networks/classification/quantize_example.py:85]: `not None` is always True.
If you don't modify `activation_quantizer_kwargs` you don't need this trick

Comment (oatawa1) [reply blueoil/networks/classification/quantize_example.py]: Thank you four raise this up. You're right, it's always true. Surprisingly, I didn't notice them.  

This happened to occur in another network code, i.e., `darknet.py`, `lm_fyolo.py`, and `yolo_v2_quantize.py`. <- I think I will fix them in other PR.

**Comment by oatawa1** [comment]
@tsawada Thanks for reviews.  I revised them. Please review it again.

File: blueoil/networks/classification/quantize_example.py [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+            activation_quantizer (callable): activation quantizer
+            activation_quantizer_kwargs(dict): Initialize kwargs for activation quantizer.
+        """
+
+        super().__init__(*args, **kwargs)
+
+        self.quantize_first_convolution = quantize_first_convolution
+        self.quantize_last_convolution = quantize_last_convolution
+
+        activation_quantizer_kwargs = activation_quantizer_kwargs if activation_quantizer_kwargs is not None else {}
```
Comment (tsawada) [inline blueoil/networks/classification/quantize_example.py:75]: I think we don't need this `a if a is not None else {}` trick at all here.  This is only necessary when we modify given args.
Here, we don't modify `activation_quantizer_kwargs` thus we can simply use `{}` as the default arg.

**Comment by oatawa1** [comment]
@tsawada Thank you.  I changed the default kwargs to {}.

Anyway, these changes will make the default parameters of these two variables inconsistent with all other networks. I wonder when we should keep it simple and when we should keep consistency with others. This is not clear to me.

**Comment by tsawada** [comment]
what about modifying #974 the same way and merging it first?

**Comment by oatawa1** [comment]
@tsawada I modified other networks in #974 and now it is merged. Please review this PR again.

**Comment by oatawa1** [comment]
@tsawada Thank you.

**Pull request merged**